pkg/daemon/ceph/client/upgrade.go

```diff
@@ -104,6 +147,183 @@ func EnableNautilusOSD(context *clusterd.Context) error {
+				// now looping over a single element since we can't address the key directly (we don't know its name)
+				for _, monCount := range versions.Mon {
+					if monCount < 3 {
+						logger.Infof("the cluster has less than 3 monitors, not performing upgrade check, running in best-effort")
+						return nil
+					}
+				}
+			}
+		// Trying to handle the case where a **single** osd is deployed and an upgrade is called
+		case "osd":
+			// if len(versions.Osd) > 1, we have different Ceph versions for some osd(s)
+			// this is fine running the upgrade we can run the upgrade checks
+			if len(versions.Osd) == 1 {
+				// now trying to parse and find how many mons are presents
+				// if we have less than 3 osds we skip the check and do best-effort
+				for _, osdCount := range versions.Osd {
+					if osdCount < 3 {
+						logger.Infof("the cluster has less than 3 OSDs, not performing upgrade check, running in best-effort")
+						return nil
+					}
+				}
+			}
+		}
+		// we don't implement any checks for mon, rgw and rbdmirror since:
+		//  - mon: the is done in the monitor code since it ensures all the mons are always in quorum before continuing
+		//  - rgw: the pod spec has a liveness probe so if the pod successfully start
+		//  - rbdmirror: you can chain as many as you want like mdss but there is no ok-to-stop logic yet
+		err = okToStopDaemon(context, deployment)
+		if err != nil {
+			return fmt.Errorf("failed to check if %s was ok to %s. %+v", deployment, action, err)
+		}
+
+	case "continue":
+		// the mon case is handled directly in the deployment where the mon checks for quorum
+		switch daemonName {
+		case "osd":
+			err := okToContinueOSDDaemon(context, namespace)
+			if err != nil {
+				return fmt.Errorf("failed to check if %s was ok to %s. %+v", deployment, action, err)
+			}
+		case "mds":
+			err := okToContinueMDSDaemon(context, namespace, deployment)
+			if err != nil {
+				return fmt.Errorf("failed to check if %s was ok to %s. %+v", deployment, action, err)
+			}
+		}
+	}
 
 	return nil
 }
+
+func okToStopDaemon(context *clusterd.Context, deployment string) error {
+	daemonID := findDaemonID(deployment)
+	daemonName, err := findDaemonName(deployment)
+	if err != nil {
+		return fmt.Errorf("%+v", err)
+	}
+
+	if !stringInSlice(daemonName, daemonNoCheck) {
+		output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", daemonName, "ok-to-stop", daemonID)
+		if err != nil {
+			return fmt.Errorf("deployment %s cannot be stopped. %+v", deployment, err)
+		}
+		logger.Infof("deployment %s is ok to be updated. %s", deployment, output)
+	}
+
+	logger.Infof("deployment %s is ok to be updated.", deployment)
+
+	return nil
+}
+
+// okToContinueOSDDaemon determines whether it's fine to go to the next osd during an upgrade
+// This basically makes sure all the PGs have settled
+func okToContinueOSDDaemon(context *clusterd.Context, namespace string) error {
+	err := util.Retry(3000, 15*time.Second, func() error {
```

Comment:
Yes it will... That's difficult, it's hard to tell how long we should wait for. I can try to do something smarter like checking for Ceph flags before doing anything on OSD.

The thing is that it's hard (with the current implementation) to refuse and/or skip an upgrade for whatever reason (recovery in progress etc). If you have any ideas on how to have better control let me know.

We would likely need to refuse an upgrade on the mons before doing anything, but again if we cancel it, how do we resume it? Thanks!

Reply:
As discussed in a call, my concerns are that we don't want to leave rook in an inconsistent state where we can't complete an orchestration. If an orchestration fails to complete or hangs for a really long time, it has other side effects on the cluster. I want to be clear on the side effects of the new upgrade requirements. 

With the current upgrade implementation where rook is very aggressive to update all the daemons without waiting very much for health, users haven't really complained about it. Obviously we want to improve it and avoid downtime where possible. However, I don't want to swing too far to the other side where orchestrations are blocked. It feels like we should have some reasonable timeout where we proceed to upgrade a daemon even if the PGs are not clean. 

@liewegas What is your perspective on how aggressive is reasonable during upgrades?

Reply:
I've changed that to 10 min since PG settling after an OSD restart shouldn't take too long to reconcile. Still I'd like @liewegas's or @jdurgin on this.